src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1454,3 +1455,101 @@ int32_t SystemNative_ReadProcessStatusInfo(pid_t pid, ProcessStatus* processStat
     return -1;
 #endif // __sun
 }
+
+int32_t SystemNative_PRead(intptr_t fd, void* buffer, int32_t bufferSize, int64_t fileOffset)
+{
+    assert(buffer != NULL);
+    assert(bufferSize >= 0);
+
+    ssize_t count;
+    while ((count = pread(ToFileDescriptor(fd), buffer, (uint32_t)bufferSize, (off_t)fileOffset)) < 0 && errno == EINTR);
+
+    assert(count >= -1 && count <= bufferSize);
+    return (int32_t)count;
+}
+
+int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64_t fileOffset)
+{
+    assert(buffer != NULL);
+    assert(bufferSize >= 0);
+
+    ssize_t count;
+    while ((count = pwrite(ToFileDescriptor(fd), buffer, (uint32_t)bufferSize, (off_t)fileOffset)) < 0 && errno == EINTR);
+
+    assert(count >= -1 && count <= bufferSize);
+    return (int32_t)count;
+}
+
+int64_t SystemNative_PReadV(intptr_t fd, struct iovec* vectors, int32_t vectorCount, int64_t fileOffset)
+{
+    assert(vectors != NULL);
+    assert(vectorCount >= 0);
+
+    ssize_t count = 0;
+    int fileDescriptor = ToFileDescriptor(fd);
+#if HAVE_PREADV
+    while ((count = preadv(fileDescriptor, vectors, (int)vectorCount, (off_t)fileOffset)) < 0 && errno == EINTR);
+#else
+    ssize_t current;
+    for (int i = 0; i < vectorCount; i++)
+    {
+        struct iovec vector = vectors[i];
+        while ((current = pread(fileDescriptor, vector.iov_base, vector.iov_len, (off_t)(fileOffset + count))) < 0 && errno == EINTR);
+
+        if (current < 0)
+        {
+            // if previous calls were succesfull, we return what we got so far
+            // otherwise, we return the error code
+            return count > 0 ? count : current;
+        }
+
+        count += current;
+
+        // we stop on the first incomplete operation (the next call would fail)
```

Comment:
The `pread` [docs](https://man7.org/linux/man-pages/man2/pwrite.2.html) are not very clear about it:

> 
       Note that it is not an error for a successful call to transfer
       fewer bytes than requested (see read(2) and write(2)).

But `read` [docs](https://man7.org/linux/man-pages/man2/read.2.html) contain more info:

> 
       It is not an error if this number is smaller than the number of
       bytes requested; this may happen for example because fewer bytes
       are actually available right now (maybe because we were close to
       end-of-file, or because we are reading from a pipe, or from a
       terminal), or because read() was interrupted by a signal.  See
       also NOTES.

I know that these are just examples and the spec does not state that 100% clear but it seems that all implementations:

* return fewer bytes on EOF (and we also want to stop here)
* when reading from a pipe and no more data available (not a case here as this method is offset based and it won't be used for pipes)
* when read was interrupted by a signal (same applies to `preadv` which we try to mimic here)

So to mimic preadv, we stop on the first incomplete operation.

I've updated the comments in the source code. @stephentoub Does that answer your quesiton?

Reply:
It answers my question, but I think the comment is still wrong.  Even, for example, in the case of EOF, the next call could succeed if the file length was extended in the interim.  I'm not suggesting the logic change, but I think the comment should.

Reply:
Good point, I am going to update the comment